encoder/vaapiencoder_vp9.cpp

```diff
@@ -191,23 +192,46 @@ bool VaapiEncoderVP9::fill(VAEncPictureParameterBufferVP9* picParam,
             picParam->reference_frames[i] = VA_INVALID_SURFACE;
     }
     else {
-        picParam->refresh_frame_flags = 0x01; // refresh last frame
+
         picParam->pic_flags.bits.frame_type = kInterFrame;
 
         ReferenceQueue::const_iterator it = m_reference.begin();
 
-        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++) {
             picParam->reference_frames[i] = (*it)->getID();
+            DEBUG("reference frame[%d] 0x%x", i, (*it)->getID());
+        }
 
+        picParam->pic_flags.bits.frame_context_idx = 0;
         // last/golden/alt is used as reference frame. L0 forward
         picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
 
         // golden and alt are last KeyFrame
         // last is last decoded frame
-        picParam->ref_flags.bits.ref_last_idx = 0;
-        picParam->ref_flags.bits.ref_gf_idx = 1;
-        picParam->ref_flags.bits.ref_arf_idx = 2;
-        picParam->pic_flags.bits.frame_context_idx = 0;
+
+        if (getReferenceMode()) {
+            // intel driver updates a new slot with every new frame and keeps
+            // the reference frames in a circular buffer, the buffer is defined
+            // with 8 slots but it is up to the application to use as many as
+            // desired. This scheme implements the use of 3 slots for
+            // last/gold/alt refs
+
+            picParam->refresh_frame_flags = 1 << m_referenceIndex[1];
+            // assign the references on only 3 slots
+            picParam->ref_flags.bits.ref_arf_idx = m_referenceIndex[1];
```

Comment:
On Mon, Sep 12, 2016 at 6:47 PM, Xu Guangxin <notifications@github.com>
wrote:

> In encoder/vaapiencoder_vp9.cpp
> <https://github.com/01org/libyami/pull/579#discussion_r78485031>:
>
> >          // last/golden/alt is used as reference frame. L0 forward
> >          picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
> >
> >          // golden and alt are last KeyFrame
> >          // last is last decoded frame
> > -        picParam->ref_flags.bits.ref_last_idx = 0;
> > -        picParam->ref_flags.bits.ref_gf_idx = 1;
> > -        picParam->ref_flags.bits.ref_arf_idx = 2;
> > -        picParam->pic_flags.bits.frame_context_idx = 0;
> > +
> > +        if (getReferenceMode()) {
> > +            // intel driver updates a new slot with every new frame and keeps
> > +            // the reference frames in a circular buffer, the buffer is defined
> > +            // with 8 slots but it is up to the application to use as many as
> > +            // desired. This scheme implements the use of 3 slots for
> > +            // last/gold/alt refs
> > +
> > +            picParam->refresh_frame_flags = 1 << m_referenceIndex[1];
> > +            // assign the references on only 3 slots
> > +            picParam->ref_flags.bits.ref_arf_idx = m_referenceIndex[1];
>
> L221-L223 is simliar to L231-L233. Could you merge this up? If we always
> treat ref_last_idx as m_referenceIndex[0], ref_gf_idx as
> m_referenceIndex[1], ref_arf_idx as m_referenceIndex[2].
> It should simplify our code. We can use following code in any case
>
> picParam->ref_flags.bits.ref_last_idx = m_referenceIndex[0];
> picParam->ref_flags.bits.ref_gf_idx = m_referenceIndex[1];
> picParam->ref_flags.bits.ref_arf_idx` = m_referenceIndex[2];
>
>
Not quite,  On this second scheme the first inter frame has to be inserted
on position 1, the next interframe on position 2 and the next wraps around
to position 0.  First scheme updates only position 0.

>  Moreover, I think the if (getReferenceMode()), can be removed. Base on
> following reason:
> 1. The m_referenceIndex and refresh_frame_flags totally decide by the
> order of m_reference. Therefore, we can handle it at referenceListUpdate.
> 2. This make code more clear, only one place need handle the
> getReferenceMode is referenceListUpdate
> What do you think?
>

getReferenceMode is a user command (this has to be fixed in a better way by
overriding setParameters from the encoder_base class).  It helps to pick
the reference scheme and it affects how m_reference and m_referenceIndex
will behave.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/01org/libyami/pull/579/files/21f72e5e3c7971528296d41d7383f0954f68b772#r78485031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEOX5t4qJgOChyo7YtUY7oLCokbFqiUks5qpgCzgaJpZM4J7IGf>
> .
>